src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/TraversingBlockBalancedTreeReader.java

```diff
@@ -0,0 +1,314 @@
+            nodeID = 1;
+            level = 1;
+            splitPackedValueStack[level] = new byte[packedBytesLength];
+            leafBlockFPStack = new long[treeDepth + 1];
+            leftNodePositions = new int[treeDepth + 1];
+            rightNodePositions = new int[treeDepth + 1];
+            splitValuesStack = new byte[treeDepth + 1][];
+            negativeDeltas = new boolean[treeDepth + 1];
+
+            in = new ByteArrayDataInput(packedIndex);
+            splitValuesStack[0] = new byte[packedBytesLength];
+            readNodeData(false);
+            scratch = new BytesRef();
+            scratch.length = bytesPerValue;
+        }
+
+        public void pushLeft()
+        {
+            int nodePosition = leftNodePositions[level];
+            nodeID *= 2;
+            level++;
+            if (splitPackedValueStack[level] == null)
+                splitPackedValueStack[level] = new byte[packedBytesLength];
+            System.arraycopy(negativeDeltas, level - 1, negativeDeltas, level, 1);
+            assert !leafNode;
+            negativeDeltas[level] = true;
+            in.setPosition(nodePosition);
+            readNodeData(true);
+        }
+
+        public void pushRight()
+        {
+            int nodePosition = rightNodePositions[level];
+            nodeID = nodeID * 2 + 1;
+            level++;
+            if (splitPackedValueStack[level] == null)
+                splitPackedValueStack[level] = new byte[packedBytesLength];
+            System.arraycopy(negativeDeltas, level - 1, negativeDeltas, level, 1);
+            assert !leafNode;
+            negativeDeltas[level] = false;
+            in.setPosition(nodePosition);
+            readNodeData(false);
+        }
+
+        public void pop()
+        {
+            nodeID /= 2;
+            level--;
+            leafNode = false;
+        }
+
+        public boolean isLeafNode()
+        {
+            return nodeID >= numLeaves;
+        }
+
+        public boolean nodeExists()
+        {
+            return nodeID - numLeaves < numLeaves;
+        }
+
+        public int getNodeID()
+        {
+            return nodeID;
+        }
+
+        public byte[] getSplitPackedValue()
+        {
+            assert !isLeafNode();
+            assert splitPackedValueStack[level] != null : "level=" + level;
+            return splitPackedValueStack[level];
+        }
+
+        public long getLeafBlockFP()
+        {
+            assert isLeafNode() : "nodeID=" + nodeID + " is not a leaf";
+            return leafBlockFPStack[level];
+        }
+
+        public BytesRef getSplitDimValue()
+        {
+            assert !isLeafNode();
+            scratch.bytes = splitValuesStack[level];
+            scratch.offset = 0;
+            return scratch;
+        }
+
+        private void readNodeData(boolean isLeft)
+        {
+            leafBlockFPStack[level] = leafBlockFPStack[level - 1];
+
+            // read leaf block FP delta
+            if (!isLeft)
```

Comment:
The pointer is already in the right spot for left children?

Reply:
This is a stack and is loaded with the file pointer for the next leaf at that level. This could be a left node or a right node but the stack is initialised in the `TraversalState` constructor with `isLeft = false`, so initially, the stack is set with the file pointer for the first left leaf in the tree. 

This is slightly confusing and I can't find a good way to document it. It is the responsibility of the writer to pack the index such that the file pointer deltas are read correctly as the tree is traversed.